.circleci/config.yml

```diff
@@ -101,25 +121,45 @@ jobs:
             done
           background: true
 
+      # Run the unit tests themselves
       - run:
           name: run tests
           command: |
-            export DISPLAY=:0
+            export DISPLAY=:0  # we need an X11 display for tests of environment rendering
+            # --cov specifies which directories to report code coverage for
+            # Since we test the installed `seals`, our source files live in `venv`, not in `src/seals`.
+            # --junitxml records test results in JUnit format. CircleCI can parse this to pretty-print results.
+            # --shard-id and --num-shards are used to split tests across parallel executors using `pytest-shard`.
+            # -n uses `pytest-xdist` to parallelize tests within a single instance.
             pytest --cov=/venv/lib/python3.7/site-packages/seals --cov=tests \
                    --junitxml=/tmp/test-reports/junit.xml \
                    --shard-id=${CIRCLE_NODE_INDEX} --num-shards=${CIRCLE_NODE_TOTAL} \
                     -n ${NUM_CPUS} -vv tests/
+            # Following two lines rewrite paths from venv/ to src/, based on `coverage:paths` in `setup.cfg`
+            # This is needed to avoid confusing Codecov
             mv .coverage .coverage.bench
-            coverage combine  # rewrite paths from virtualenv to src/
+            coverage combine
       - codecov/upload
 
+      # Upload the test results and resource usage to CircleCI
       - store_artifacts:
           path: /tmp/test-reports
           destination: test-reports
+      - store_artifacts:
+          path: /tmp/resource-usage
+          destination: resource-usage
+      # store_test_results uploads the files and tells CircleCI that it should parse them as test results
       - store_test_results:
           path: /tmp/test-reports
           unit-test:
 
+# Workflows define the jobs to actually run on CircleCI.
+# Here we have just a single workflow, `test`.
+# We attach two contexts to both jobs:
```

Comment:
```suggestion
# We attach two contexts to both workflows:
```

Name overlapping with `jobs` abstraction from above

Reply:
Hm. The contexts are associated with the workflow-job pair I guess, but I view it more closely tied to the jobs. You can have different jobs in the same workflow with different contexts, and the same job in different workflows with different contexts.

Reply:
Oh I see. I didn't really understand the relationship between jobs and contexts from this brief comment, but I think it's reasonable that someone should be reading the documentation to understand what's going on at a deeper level.